### PR TITLE
feat: add replaceMediaStream to session

### PR DESCRIPTION
**Pull request by Krivega**

### CODE REVIEW

**Comment by jmillan** [comment]
Hi,

I would rather don't renegotiate inside `replaceMediaStream` and let it just replace the media stream.

File: lib/RTCSession.js [hunk 1]
```diff
@@ -1344,6 +1342,38 @@ module.exports = class RTCSession extends EventEmitter
...
 
+  replaceMediaStream(mediaStream, options)
+  {
+    debug('replaceMediaStream()');
+
+    return new Promise((resolve, reject) =>
+    {
+      this._removeMediaStream(this._localMediaStream);
+      this._addMediaStream(mediaStream);
+      this.renegotiate(options, resolve, reject);
```
Comment (jmillan) [inline lib/RTCSession.js:1353]: call to `renegotiate` should go out of this method.

File: lib/RTCSession.js [hunk 2]
```diff
@@ -1344,6 +1342,38 @@ module.exports = class RTCSession extends EventEmitter
...
+    });
+  }
+
+
+  _addMediaStream(stream)
+  {
+    stream.getTracks().forEach((track) => this._connection.addTrack(track, stream));
```
Comment (jmillan) [inline lib/RTCSession.js:1360]: Please respect the current style, use braces as in the rest of the code.

File: lib/RTCSession.js [hunk 3]
```diff
@@ -1344,6 +1342,38 @@ module.exports = class RTCSession extends EventEmitter
...
+
+  _removeMediaStream(stream)
+  {
+    this._connection.getSenders().forEach((sender) =>
+    {
+      stream.getTracks().forEach((track) =>
+      {
+        if (sender.track === track)
+        {
+          this._connection.removeTrack(sender);
```
Comment (jmillan) [inline lib/RTCSession.js:1371]: `break` here.

File: lib/RTCSession.js [hunk 4]
```diff
@@ -1821,7 +1851,16 @@ module.exports = class RTCSession extends EventEmitter
 
   _createRTCConnection(pcConfig, rtcConstraints)
   {
-    this._connection = new RTCPeerConnection(pcConfig, rtcConstraints);
+    this._connection = new RTCPeerConnection(
```
Comment (jmillan) [inline lib/RTCSession.js:1854]: This change has nothing to do here.

Comment (Krivega) [reply lib/RTCSession.js:1854]: also surprised

**Comment by Krivega** [comment]
> Hi,
> 
> I would rather don't renegotiate inside `replaceMediaStream` and let it just replace the media stream.

fail to renegotiate leave?

File: lib/RTCSession.js [hunk 5]
```diff
@@ -1290,7 +1287,7 @@ module.exports = class RTCSession extends EventEmitter
     return true;
   }
 
-  renegotiate(options = {}, done)
+  renegotiate(options = {}, done, fail)
```
Comment (jmillan) [inline lib/RTCSession.js:1290]: `fail` needs to go away. Failures are handled by terminating the session.

File: lib/RTCSession.js (hunk 3 above)
Comment (Krivega) [reply lib/RTCSession.js:1370]: why do you need to break?

Comment (jmillan) [reply lib/RTCSession.js:1369]: There is one track for sender. If you found it then there's no reason to keep iterating.

Comment (jmillan) [reply lib/RTCSession.js:1369]: wait, we are inside a `forEach`

Comment (Krivega) [reply lib/RTCSession.js:1369]: yes, it needs to be rewritten to `for()`

Comment (jmillan) [reply lib/RTCSession.js:1369]:
```js
if (stream.getTracks().includes(sender.track) {
  this._connection.removeTrack(sender);
}
```

Comment (jmillan) [reply lib/RTCSession.js:1369]: no need for the nested `forEach`

Comment (Krivega) [reply lib/RTCSession.js:1369]: I thought about
```js
    stream.getTracks().forEach((track) => {
      const senderMediaStream = this._connection.getSenders().find((sender) => sender.track === track);

      if (senderMediaStream) {
        this._connection.removeTrack(senderMediaStream);
      }
    });
``` but it's better

Comment (Krivega) [reply lib/RTCSession.js]: or 
```js

    this._connection.getSenders()
    .filter((sender) => stream.getTracks().includes(sender.track))
    .forEach((sender) => { 
      this._connection.removeTrack(sender);
    });
```

**Comment by ibc** [comment]
Honestly I don't like this PR much. We don't need an API to replace the current `MediaStream`. The application can get the `RTCPeerConnection` via `session.connection` and then use the WebRTC API to replace the audio or the video track using (for example):

```js
const rtpSender = session.connection.getSenders()[0];

await rtpSender.replaceTrack(newAudioTrack)`.
```

So thanks for this effort but I'm against adding sugar API to JsSIP (which means lot of maintenance for us).

**Comment by jmillan** [comment]
The negotiation has been removed from the PR. I asked for that. There has been some other cleanup.

I had the same feeling @ibc, but if you look at the changes now it just does that; remove tracks from current stream and add tracks from a new one. It's a single new public method,  syntactic sugar but its so harmless, and replacing the stream is something for which actually we get questions about in the mailing list.

Please review the changes again.

**Comment by ibc** [comment]
> syntactic sugar but its so harmless, and replacing the stream is something for which actually we get questions about in the mailing list.

Why do we assume that we just handle a single `MediaStream`? We may handle an audio track from stream1 and video track from stream2.

**Comment by ibc** [comment]
> replacing the stream is something for which actually we get questions about in the mailing list

Because people do not take a look to the WebRTC API and avoid using `session.connection`. That does not mean that we should provide a replacement for them.

But ok, changes in this PR look much tiny now. However, TypeScript required.

**Comment by KraftyKraft** [comment]
> @KraftyKraft can you also send a PR to https://github.com/versatica/jssip.net with the new API?

Hi! Did you mean to mention @Krivega ? :)

**Comment by ibc** [comment]
Sure, crazy GitHub nick suggestions :)

**Comment by Krivega** [comment]
sure

**Comment by ibc** [comment]
> can you also send a PR to https://github.com/versatica/jssip.net with the new API?

Hi @Krivega, waiting for this in order to accept the PR :)

**Comment by Krivega** [comment]
@ibc with `replaceTrack` switching the track should not require negotiation
but I have doubts about adding a track, to be consistent, you need to check tracks that we did not replace and delete. Or, at the beginning, check the current tracks and compare with those that come and in case of mismatch give an error / warning

**Comment by ibc** [comment]
> @ibc with `replaceTrack` switching the track should not require negotiation
> but I have doubts about adding a track, to be consistent, you need to check tracks that we did not replace and delete. Or, at the beginning, check the current tracks and compare with those that come and in case of mismatch give an error / warning

That is exactly one of the things I just don't want to do within JsSIP internals. If the app adds a new track it should renegotiate. It the app replaces a track, it shouldn't.

**Comment by jmillan** [comment]
Guys,

This was a great attempt to create a higher level public method for a common usage, but it has demonstrated to break the paradigm of the library adding a complex to comprehend and maintain method that would require just few lines of code in de the app.

@Krivega, thanks for the efforts, I hope you understand the decision to close the PR. :+1:

**Pull request closed without merging**